### ISSUE TITLE
[express-oauth2-jwt-bearer] What is the best practice to allow request to go through despite token being unauthorized?

### ISSUE THREAD

**Issue by nbkhope** [auth0/node-oauth2-jwt-bearer#92]
I'm migrating to `express-oauth2-jwt-bearer` and trying to fit into my existing project authentication & authorization pipeline.

I would like to have `auth()` try to authenticate, but still let the user through in an unauthenticated state. Currently, the middleware [calls next with the error](https://github.com/auth0/node-oauth2-jwt-bearer/blob/fe0dace80c34a35733542f5df7783dd996f7297e/packages/express-oauth2-jwt-bearer/src/index.ts#L77) if unauthorized, which by default goes to the Express error handling middleware and returns a 401 response.

I could add a second step in the pipeline just to catch that possible error, but then that error middleware would also catch any other kind of error and would get in the way of my standard error pipeline.

I am used to `passport` and this the way I would handle it before:

```
passport.authenticate('jwt', function(err, user, info) {
  // here I would do whatever in case of invalid token, letting user through
```

**Comment by adamjmcgrath** [comment]
> I could add a second step in the pipeline just to catch that possible error, but then that error middleware would also catch any other kind of error and would get in the way of my standard error pipeline.

You could also just do

```js
// untested:
const { auth } = require('express-oauth2-jwt-bearer');

const authNoError = (config) => {
  const mw = auth(config);
  return (req, res, next) => {
    mw(req, res, () => next())
  }
}

app.use(authNoError());
```

**Comment by nbkhope** [comment]
Thanks for the response, @adamjmcgrath.

One way I got it to work was to add an extra error handler, then rethrow the error if it does not match the error name, header, status 401, etc. Still don't like that though.

As for your first proposed solution, that could work, but I'm afraid of cases such as expired tokens going through. In that case, there would still be an Authorization header set, although the value invalid. I would like the response to still go through, as though you were anonymous user. Is that unreasonable?

As for your second proposed solution, I have to try that out and get to understand the `() => next()` part (at first glancing assuming it would just get rid of the error and letting it through; that could work).

> Does passport have the option to throw/not throw based on something similar? Which Passport strategy are you referring to?

The example I used uses [`passport-jwt`](https://www.npmjs.com/package/passport-jwt), but the actual error handling mechanism that lets me keep the pipeline going is from `passport` itself, the callback parameter to `authenticate()`:

https://github.com/jaredhanson/passport/blob/cfdbd4a762b51e339ebfea931d65bccbbde53282/lib/middleware/authenticate.js#L34-L42
